crates/js-backend/src/function.rs

```diff
@@ -0,0 +1,280 @@
+    pub(crate) js_func: js_sys::Function,
+    pub(crate) sig: FuncSig,
+}
+
+impl StoredFunction {
+    pub(crate) fn new(js_func: js_sys::Function, sig: FuncSig) -> Self {
+        Self { js_func, sig }
+    }
+}
+
+impl WasmExportFunction {
+    pub(crate) fn new_stored(
+        ctx: &mut impl AsContextMut<JsWasmBackend>,
+        instance: JsInstance,
+        func: js_sys::Function,
+        sig: FuncSig,
+    ) -> Self {
+        let handle = ctx
+            .as_context_mut()
+            .inner
+            .store_function(StoredFunction::new(func, sig));
+
+        Self {
+            store_handle: handle,
+            bound_instance: instance,
+        }
+    }
+
+    pub(crate) fn stored_mut<'store>(
+        &self,
+        ctx: JsContextMut<'store>,
+    ) -> &'store mut StoredFunction {
+        &mut ctx.inner.functions[self.store_handle]
+    }
+
+    fn call_inner(
+        &self,
+        store: &mut impl AsContextMut<JsWasmBackend>,
+        args: &[WValue],
+    ) -> RuntimeResult<Vec<WValue>> {
+        let params = js_array_from_wval_array(args);
+        let stored_func = self.stored_mut(store.as_context_mut());
+        let result = js_sys::Reflect::apply(&stored_func.js_func, &JsValue::NULL, &params)
+            .map_err(|e| {
+                web_sys::console::log_2(&"failed to apply func".into(), &e);
+                RuntimeError::Other(anyhow!("Failed to apply func"))
+            })?;
+
+        let result_types = stored_func.sig.returns();
+        match result_types.len() {
+            0 => Ok(vec![]),
+            1 => {
+                // Single value returned as is.
+                let value = wval_from_js(&result_types[0], &result);
+                Ok(vec![value])
+            }
+            results_number => {
+                // Multiple return values are returned as JS array of values.
+                let result_array: Array = result.into();
+                if result_array.length() as usize != results_number {
+                    Err(RuntimeError::IncorrectResultsNumber {
+                        expected: results_number,
+                        actual: result_array.length() as usize,
+                    })
+                } else {
+                    Ok(wval_array_from_js_array(&result_array, result_types.iter()))
+                }
+            }
+        }
+    }
+}
+
+impl HostImportFunction {
+    pub(crate) fn stored<'store>(&self, ctx: &JsContext<'store>) -> &'store StoredFunction {
+        &ctx.inner.functions[self.store_handle]
+    }
+
+    pub(crate) fn stored_mut<'store>(
+        &self,
+        ctx: JsContextMut<'store>,
+    ) -> &'store mut StoredFunction {
+        &mut ctx.inner.functions[self.store_handle]
+    }
+}
+
+// Safety: this is safe because its intended to run in single thread
+unsafe impl Send for HostImportFunction {}
+unsafe impl Sync for HostImportFunction {}
+unsafe impl Send for WasmExportFunction {}
+unsafe impl Sync for WasmExportFunction {}
```

Comment:
I think that these definitions are so important that it's better to move it to the top of the file after uses